sklearn/compose/_column_transformer.py

```diff
@@ -638,10 +639,29 @@ def _hstack(self, Xs):
         This allows subclasses to control the stacking behavior, while reusing
         everything else from ColumnTransformer.
 
+        If Xs are all DataFrames and their indexes are the same then a
+        DataFrame is returned.
+
         Parameters
         ----------
         Xs : list of {array-like, sparse matrix, dataframe}
         """
+        # Check that all output from transformers are pandas DataFrames
+        if all(hasattr(X, "iloc") for X in Xs):
+            if any(not Xs[0].index.equals(X.index) for X in Xs[1:]):
+                warnings.warn(
+                    "The DataFrame's indexes for each transformer do not "
+                    "match. A numpy array or a sparse matrix will be "
+                    "returned.")
+                return self._hstack_np(Xs)
+
+            import pandas as pd
+            return pd.concat(Xs, axis=1)
+
+        return self._hstack_np(Xs)
+
+    def _hstack_np(self, Xs):
```

Comment:
To be a purist, sparse is not really `np` but coming from `scipy` :)

I am wondering if we should have 2 functions here. I would prefer to have an if/else statement in the `_hstack`.

Reply:
In the `_safe_indexing` we indeed have 3 functions: `_array_indexing`, `_pandas_indexing`, and `_list_indexing`. We could indeed have something similar then.